staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2181 @@
+        //
+        // Vcan can not be configured as CANFD interface, but it is
+        // possible to configure its MTU to 64 bytes. So if a messages
+        // bigger than 8 bytes is being received we consider it as
+        // CANFD message.
+        let can_name = self.controller.read().unwrap().can_name.clone();
+        if self.check_features(VIRTIO_CAN_F_CAN_FD)
+            && response.length.to_native() > 8
+            && can_name == "vcan0"
+        {
+            response.flags = (response.flags.to_native() | CAN_FRMF_TYPE_FD).into();
+            warn!("\n\n\nCANFD VCAN0\n\n");
+        }
+
+        // Adapt CAN length based on negotiated features
+        if (response.flags.to_native() & CAN_FRMF_TYPE_FD) != 0 {
+            can_rx.flags = (can_rx.flags.to_native() | VIRTIO_CAN_FLAGS_FD).into();
+            if response.length.to_native() > 64 {
+                warn!(
+                    "%s(): Cut length from {} to 64\n",
+                    response.length.to_native()
+                );
+                can_rx.length = 64.into();
+            }
+        } else if response.length.to_native() > 8 {
+            warn!(
+                "%s(): Cut length from {} to 8\n",
+                response.length.to_native()
+            );
+            can_rx.length = 8.into();
+        }
+
+        can_rx.sdu.copy_from_slice(&response.sdu[0..64]);
+        CanController::print_can_frame(can_rx);
+
+        Ok(can_rx)
+    }
+
+    fn process_ctrl_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_ctrl_requests");
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let atomic_mem = self.mem.as_ref().unwrap().memory();
+
+            let mut reader = desc_chain
+                .clone()
+                .reader(&atomic_mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let mut writer = desc_chain
+                .clone()
+                .writer(&atomic_mem)
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            let request = reader
+                .read_obj::<VirtioCanCtrlRequest>()
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            // This implementation requires the CAN devices to be already in UP state
+            // before starting. This code does not trigger state changes [UP/DOWN] of
+            // the host CAN devices.
+            let response = match request.msg_type.into() {
+                VIRTIO_CAN_SET_CTRL_MODE_START => {
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_START");
+                    if self.controller.write().unwrap().ctrl_state == CAN_CS_STARTED {
+                        Ok(VIRTIO_CAN_RESULT_NOT_OK)
+                    } else {
+                        self.controller.write().unwrap().busoff = false;
```

Comment:
Is it OK to change the `busoff` here? The spec states
```
The bit VIRTIO_CAN_S_CTRL_BUSOFF in status is used to indicate the unsolicited CAN controller state
change from started to stopped due to a detected bus off condition.
```
So I think the switching to `False` shall happen independent of the driver although I do not know when or how.

Reply:
Since that was similar to what you wrote below, I tried to reply to all of them together in the next thread.